kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java

```diff
@@ -0,0 +1,291 @@
+import io.streamnative.pulsar.handlers.kop.coordinator.group.GroupMetadata;
+import io.streamnative.pulsar.handlers.kop.utils.KopTopic;
+import io.streamnative.pulsar.handlers.kop.utils.ShutdownableThread;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.pulsar.broker.loadbalance.LoadManager;
+import org.apache.pulsar.common.util.Murmur3_32Hash;
+import org.apache.pulsar.metadata.api.GetResult;
+import org.apache.pulsar.metadata.api.MetadataStore;
+import org.apache.pulsar.metadata.api.Notification;
+
+@Slf4j
+public class KopEventManager {
+    private static final String REGEX = "^(.*)://\\[?([0-9a-zA-Z\\-%._:]*)\\]?:(-?[0-9]+)";
+    private static final Pattern PATTERN = Pattern.compile(REGEX);
+
+    private static final String kopEventThreadName = "kop-event-thread";
+    private final ReentrantLock putLock = new ReentrantLock();
+    private static final LinkedBlockingQueue<KopEvent> queue =
+            new LinkedBlockingQueue<>();
+    private final KopEventThread thread =
+            new KopEventThread(kopEventThreadName);
+    private final GroupCoordinator coordinator;
+    private final AdminManager adminManager;
+    private final DeletionTopicsHandler deletionTopicsHandler;
+    private final BrokersChangeHandler brokersChangeHandler;
+    private final MetadataStore metadataStore;
+
+    public KopEventManager(GroupCoordinator coordinator,
+                           AdminManager adminManager,
+                           MetadataStore metadataStore) {
+        this.coordinator = coordinator;
+        this.adminManager = adminManager;
+        this.deletionTopicsHandler = new DeletionTopicsHandler(this);
+        this.brokersChangeHandler = new BrokersChangeHandler(this);
+        this.metadataStore = metadataStore;
+    }
+
+    public void start() {
+        registerChildChangeHandler();
+        thread.start();
+    }
+
+    public void close() {
+        try {
+            thread.shutdown();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            log.error("Interrupted at shutting down {}", kopEventThreadName);
+        }
+
+    }
+
+
+    public void put(KopEvent event) {
+        try {
+            putLock.lock();
+            queue.put(event);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            log.error("Error put event {} to coordinator event queue {}", event, e);
+        } finally {
+            putLock.unlock();
+        }
+    }
```

Comment:
We prefer the following lock idiom.

```java
lock.lock();
try {
    // ...
} finally {
    lock.unlock();
}
```

Because if `lock.lock()` failed with exceptions, `.unlock()` would still be called.
See examples in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html.

Reply:
The same applies to `clearAndPut`.

Reply:
good comment!